Modules/mathmodule.c

```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
     return count;
 }
 
+/* Integer square root
+
+Given a nonnegative integer `n`, we want to compute the largest integer
+`a` for which `a * a <= n`, or equivalently the integer part of the exact
+square root of `n`.
+
+We use an adaptive-precision pure-integer version of Newton's iteration. Given
+a positive integer `n`, the algorithm produces at each iteration an
+approximation `a` to the square root of `n >> s` for some even integer `s`,
+with `s` decreasing as the iterations progress. On the final iteration, `s` is
+zero and we have an approximation to the square root of `n` itself.
+
+At every step, the approximation `a` is strictly within 1.0 of the true square
+root, so we have
+
+   (a - 1)**2 < (n >> s) < (a + 1)**2
+
+After the final iteration, a check-and-correct step is needed to determine
+whether `a` or `a - 1` gives the desired integer square root of `n`.
+
+The algorithm is remarkable in its simplicity. There's no need for a
+per-iteration check-and-correct step, and termination is straightforward:
+the number of iterations is known in advance (and is roughly log2(log2(n))).
+The only tricky part of the correctness proof is in establishing that
+the bound (a - 1)**2 < (n >> s) < (a + 1)**2 is maintained from one
+step to the next. A formal, computer-verified proof of this (using Lean)
+can be found here:
+
+    https://github.com/mdickinson/snippets/blob/master/proofs/isqrt/src/isqrt.lean
+
+The comments in that file also include an informal proof. Note that the
+Python code given in comments there is recursive, but is equivalent to the
+algorithm given below.
+
+Equivalent Python code:
+
+
+    import operator
+
+    def isqrt(n):
+        """
+        Return the largest integer not exceeding the square root of the input.
+        """
+        n = operator.index(n)
+
+        if n < 0:
+            raise ValueError("isqrt() argument must be nonnegative")
+        elif n == 0:
+            return 0
+
+        c = (n.bit_length() - 1) // 2
+        d = 0
+        a = 1
+        for s in reversed(range(c.bit_length())):
+            e = d
+            d = c >> s
+            a = (a << d - e - 1) + (n >> 2*c - e - d + 1) // a
+            # Loop invariant
+            assert (a-1)**2 < n >> 2*(c - d) < (a+1)**2
+
+        return a - (a*a > n)
+
+*/
+
+/*[clinic input]
+math.isqrt
+
+    n : 'O'
+    /
+
+Return the largest integer not exceeding the square root of the input.
```

Comment:
This is a bit convoluted. "Return the integer part of the square root of the input." may be more direct and accessible.